.github/workflows/main.yml

```diff
@@ -29,67 +29,143 @@ jobs:
   linux:
     runs-on: ubuntu-latest
     needs: [check]
+    outputs:
+      output: ${{ steps.check-diff.outputs.cargo-lock }}
     container:
       image: ghcr.io/veracruz-project/veracruz/ci@sha256:b5eb834cd58c69e6acc19b1b19c6345a020ad2ad090cdab3bc9f7d5d48fb00ba
       volumes:
         - ${{ github.workspace }}:/work/veracruz
     steps:
       - name: Check out the Veracruz repository
-        uses: actions/checkout@v2
+        uses: actions/checkout@v3
         with:
           submodules: recursive
       - name: Running linux test script
         id: linux-build-and-test
         run: |
-            make -C /work/veracruz/workspaces linux-tests
+           make -C /work/veracruz/workspaces linux-tests
+      - name: Move back to veracruz root
+        run: |
+          cd /work/veracruz
+          git config --global --add safe.directory "$GITHUB_WORKSPACE"
+      - name: Check modification to Cargo.lock 
+        id: check-diff
+        run: |
+          # Find if any Cargo.lock changed, pad them into a line and trim leading and trailing whitespace.
+          file_changed=$(git diff --diff-filter=ACMUXTRD --name-only -- '**Cargo.lock' | tr '\n' ' ' | xargs)
+          echo "cargo-lock=$file_changed" >> $GITHUB_OUTPUT
+      - name: Output list of changed Cargo.lock
+        id: output-changed-cargo-lock
+        if: steps.check-diff.outputs.cargo-lock != ''
+        run: |
+            echo "::warning::Cargo.lock files modified" 
+            echo "::warning::Cargo.lock change list: ${{ steps.check-diff.outputs.cargo-lock }}" 
+      - name: Upload Cargo.lock files
+        id: upload-changed-cargo-lock
+        if: steps.check-diff.outputs.cargo-lock != ''
+        uses: actions/upload-artifact@v3
+        with: 
+          name: linux 
+          path: workspaces/**/Cargo.lock
 
   nitro:
     runs-on: ubuntu-latest
     needs: [check]
+    outputs:
+      output: ${{ steps.check-diff.outputs.cargo-lock }}
     container:
       image: ghcr.io/veracruz-project/veracruz/ci@sha256:b5eb834cd58c69e6acc19b1b19c6345a020ad2ad090cdab3bc9f7d5d48fb00ba
       volumes:
         - ${{ github.workspace }}:/work/veracruz
     steps:
       - name: Check out the Veracruz repository
-        uses: actions/checkout@v2
+        uses: actions/checkout@v3
         with:
           submodules: recursive
+      - name: add the GITHUB_WORKSPACE into git config
+        run: |
+          git config --global --add safe.directory "$GITHUB_WORKSPACE"
       - name: Running Nitro test script
         id: nitro-build
         run: |
             make -C /work/veracruz/workspaces nitro
+      - name: Check modification to Cargo.lock 
+        id: check-diff
+        run: |
+          file_changed=$(git diff --diff-filter=ACMUXTRD --name-only -- '**Cargo.lock' | tr '\n' ' ' | xargs)
```

Comment:
I don't think we need both `tr` and `xargs`

Reply:
The `GITHUB` output dislikes the existence of line-break. The last `xargs` removes starting and trailing white space.

Reply:
I am just surprised it makes any difference, because I did a few tests:
```
$ echo >> workspaces/applications/Cargo.lock
$ echo >> workspaces/data-generators/Cargo.lock
$ foo=$(git diff --diff-filter=ACMUXTRD --name-only -- '**Cargo.lock' | tr '\n' ' ' | xargs) ; echo $foo
workspaces/applications/Cargo.lock workspaces/data-generators/Cargo.lock
$ foo=$(git diff --diff-filter=ACMUXTRD --name-only -- '**Cargo.lock' | tr '\n' ' ') ; echo $foo
workspaces/applications/Cargo.lock workspaces/data-generators/Cargo.lock
$ foo=$(git diff --diff-filter=ACMUXTRD --name-only -- '**Cargo.lock' | xargs) ; echo $foo
workspaces/applications/Cargo.lock workspaces/data-generators/Cargo.lock
```
I think the trailing whitespace is being trimmed in the assignment. I will resolve the conversation.